lib/libzpool/uzfs_io.c

```diff
@@ -36,29 +35,22 @@ uzfs_write_data(zvol_state_t *zv, char *buf, uint64_t offset, uint64_t len,
 	objset_t *os = zv->zv_objset;
 	rl_t *rl, *mrl;
 	int ret = 0, error;
-	uint64_t r_offset, r_len;
-	uint64_t r_moffset, r_mlen;
+	uint64_t r_offset;
 	metaobj_blk_offset_t metablk;
 	uint64_t metadatasize = zv->zv_volmetadatasize;
 	uint64_t len_in_first_aligned_block = 0;
 
-	sync = dmu_objset_syncprop(os);
+	sync = (dmu_objset_syncprop(os) == ZFS_SYNC_ALWAYS) ? 1 : 0;
 	if (zv->zv_volmetablocksize == 0)
 		metadata = NULL;
-	/*
-	 * Taking lock on entire block at ZFS layer.
-	 * Handling the case where readlen is smaller than blocksize.
-	 * This can also be avoided later for better performance.
-	 */
-	r_offset = (offset / blocksize) * blocksize;
-	r_len = ((len + blocksize - 1) / blocksize) * blocksize;
 
+	r_offset = P2ALIGN_TYPED(offset, blocksize, uint64_t);
 	len_in_first_aligned_block = (blocksize - (offset - r_offset));
 
 	if (len_in_first_aligned_block > len)
 		len_in_first_aligned_block = len;
 
-	rl = zfs_range_lock(&zv->zv_range_lock, r_offset, r_len, RL_WRITER);
+	rl = zfs_range_lock(&zv->zv_range_lock, offset, len, RL_WRITER);
```

Comment:
shouldn't we lock r_offset, r_len? The part which is physically written to disk?

Reply:
right @jkryl .. we don't need r_offset thing and actually, upstream zfs does this way, as, zfs takes care of multiple partial writes to same block by taking locks at appropriate places. I just mentioned this in one of my above comments